File: puzzles/python3/mayan-calc/mayan_calc.py
import math
from typing import Dict, List


MayanNumeral = str

BASE = 20


def read_num(numeral_dict: Dict[MayanNumeral, int], height: int) -> List[int]:
    numeral: MayanNumeral = ""
    num_list: List[int] = []

    nb_lines = int(input())
    for i in range(nb_lines):
        num_line = input()
        numeral += num_line
        numeral += '\n'
        if i % height == height - 1:
            # last column
            number = numeral_dict[numeral]
            num_list.append(number)
            numeral = ""  # reset
    return num_list[::-1]


def convert_base20_to_10(num_list: List[int]) -> int:
    number = 0
    i = len(num_list) - 1
    while i >= 0:
        number += num_list[i] * math.pow(BASE, i)
        i -= 1
    return number


def convert_base10_to_20(number: int) -> List[int]:
    num_list: List[int] = []
    # avoid division by 0
    if number == 0:
        num_list.append(0)
        return num_list

    while number != 0:
        num_list.append(number % BASE)  # remainder
        number //= BASE

    return num_list[::-1]


if __name__ == "__main__":
    width, height = map(int, input().split())
    numerals: List[MayanNumeral] = []
    numeral_dict: Dict[MayanNumeral, int] = {}  # mayan numeral => base 20 number

    for _ in range(height):
        row: MayanNumeral = input()
        numerals.append(row)

    # Store the mayan numerals in a dictionary
    for i in range(BASE):
        numeral: MayanNumeral = ""
        column: int = i * width
        for row in numerals:
            for j in range(column, column + width):
                numeral += row[j]
            numeral += '\n'
        numeral_dict[numeral] = i

    # Read the 2 numbers
    num_list1 = read_num(numeral_dict, height)
    num_list2 = read_num(numeral_dict, height)

    num1 = convert_base20_to_10(num_list1)
    num2 = convert_base20_to_10(num_list2)

    operation = input()

    # Calculate the result of the operation in base 10
    result = 0
    if operation == '+':
        result = num1 + num2
    elif operation == '-':
        result = num1 - num2
    elif operation == '*':
        result = num1 * num2
    elif operation == '/':
        result = num1 / num2

    # Convert the result back to the original base
    num_list = convert_base10_to_20(result)
    for number in num_list:
        for numeral, n in numeral_dict.items():
            if n == number:
                print(numeral, end='')

File: puzzles/kotlin/src/mayan-calc/mayan-calc.kt
import java.util.Scanner

const val BASE = 20
val input = Scanner(System.`in`)

fun main(args : Array<String>) {
  val width = input.nextInt()
  val height = input.nextInt()
  val numerals = ArrayList<String>() // mayan numerals
  val numeralMap = HashMap<String, Long>() // mayan numeral -> base 20 number

  for (i in 0 until height) {
    val row = input.next()
    numerals.add(row)
  }

  // Store the mayan numerals in a map
  for (i in 0 until BASE) {
    val numeral = StringBuilder()
    val column = i * width

    for (row in numerals) {
      for (j in column until column + width) {
        numeral.append(row[j])
      }
      numeral.append('\n')
    }

    numeralMap[numeral.toString()] = i.toLong()
  }

  // Read the first and second number
  val numArray1 = readNum(numeralMap, height)
  val numArray2 = readNum(numeralMap, height)
  val num1 = convertBase20To10(numArray1)
  val num2 = convertBase20To10(numArray2)

  val operation = input.next()

  // Calculate the result of the operation in base 10
  val result = when (operation) {
    "+" -> num1 + num2
    "-" -> num1 - num2
    "*" -> num1 * num2
    else -> num1 / num2
  }

  // Convert the result back to the original base
  val numArray = convertBase10To20(result)
  for (number in numArray) {
    for ((numeral, n) in numeralMap) {
      if (n == number) {
        print(numeral)
      }
    }
  }
}

fun readNum(numeralMap: HashMap<String, Long>, height: Int): List<Long> {
  var numeral = StringBuilder()
  val numArray = ArrayList<Long>()

  val nbLines = input.nextInt()
  for (i in 0 until nbLines) {
    val numLine = input.next()
    numeral.append(numLine)
    numeral.append('\n')
    if (i % height == height - 1) {
      // last column
      val number = numeralMap.getOrDefault(numeral.toString(), 0L)
      numArray.add(number)
      numeral = StringBuilder() // reset
    }
  }

  return numArray.reversed()
}

fun convertBase20To10(numArray: List<Long>): Long {
  return numArray.mapIndexed { i, num -> num * Math.pow(BASE.toDouble(), i.toDouble()).toLong() }.sum()
}

fun convertBase10To20(num: Long): List<Long>  {
  var number = num
  val numArray = ArrayList<Long>()

  // avoid division by 0
  if (number == 0L) {
    numArray.add(0L)
    return numArray
  }

  while (number != 0L) {
    numArray.add(number % BASE)
    number /= BASE
  }

  return numArray.reversed()
}
